File: src/minidbms/sql_compiler/compiler.py
"""Stable Python facade over zby's C++17 SQL compiler."""
from dataclasses import dataclass
from minidbms.common.interfaces import CatalogView
from minidbms.common.plans import PlanNode
from .ast import Statement
from .lexer import Lexer
from .parser import Parser, statement_from_native
from .semantic import compile_native
from ._convert import plan


@dataclass(frozen=True)
class CompilationResult:
    ast: Statement
    bound_ast: Statement
    plan_before: PlanNode
    plan_after: PlanNode
    tokens: list
    native: dict


class SQLCompiler:
    def prepare(self, sql_text: str) -> list[Statement]:
        return Parser().parse(Lexer().tokenize(sql_text))

    def compile_statement(self, statement: Statement, catalog: CatalogView,
                          optimize: bool = True) -> CompilationResult:
        detail = compile_native(statement, catalog, optimize)
        bound = statement_from_native(detail["bound_ast"], statement.source)
        bound.native["plan_before"] = detail["plan_before"]
        bound.native["plan_after"] = detail["plan_after"]
        before = plan(detail["plan_before"])
        after = plan(detail["plan_after"]) if optimize else before
        return CompilationResult(statement, bound, before, after, detail["tokens"], detail)

    def compile_detailed(self, sql_text: str, catalog: CatalogView,
                         optimize: bool = True) -> list[CompilationResult]:
        return [self.compile_statement(s, catalog, optimize) for s in self.prepare(sql_text)]

    def compile(self, sql_text: str, catalog: CatalogView,
                optimize: bool = True) -> list[PlanNode]:
        return [d.plan_after for d in self.compile_detailed(sql_text, catalog, optimize)]

    def compile_and_execute(self, sql_text, catalog, executor, optimize=True):
        """Execute each statement before binding the next against the real Catalog."""
        results = []
        for statement in self.prepare(sql_text):
            detail = self.compile_statement(statement, catalog, optimize)
            results.append(executor.execute(detail.plan_after))
        return results

File: src/minidbms/sql_compiler/native/src/compiler.cc
#include "minisql/compiler.h"

namespace minisql {
ParseResult Compiler::prepare(const std::string& sql, bool trace) const {
  return LL1Parser().parse(tokenize(sql), trace);
}

CompilationResult Compiler::compile_statement(const StmtPtr& statement, const CatalogView& catalog, bool optimize) const {
  const auto bound = SemanticAnalyzer(catalog).analyze(statement);
  const auto before = build_plan(bound);
  return {statement, bound, before, optimize ? optimize_plan(before) : before, {}};
}

std::vector<CompilationResult> Compiler::compile_detailed(const std::string& sql, const CatalogView& catalog, bool optimize) const {
  const auto tokens = tokenize(sql);
  const auto parsed = LL1Parser().parse(tokens);
  std::vector<CompilationResult> results;
  std::size_t start = 0;
  for (const auto& statement : parsed.statements) {
    auto detail = compile_statement(statement, catalog, optimize);
    std::size_t end = start;
    while (end < tokens.size() && tokens[end].type != ";") ++end;
    if (end < tokens.size()) ++end;
    detail.tokens.assign(tokens.begin() + static_cast<std::ptrdiff_t>(start), tokens.begin() + static_cast<std::ptrdiff_t>(end));
    start = end;
    results.push_back(std::move(detail));
  }
  return results;
}

std::vector<PlanPtr> Compiler::compile(const std::string& sql, const CatalogView& catalog, bool optimize) const {
  std::vector<PlanPtr> plans;
  for (const auto& result : compile_detailed(sql, catalog, optimize)) plans.push_back(result.plan_after);
  return plans;
}

std::vector<ExecutionResult> Compiler::compile_and_execute(const std::string& sql, const CatalogView& catalog,
                                                          Executor& executor, bool optimize) const {
  const auto parsed = prepare(sql);
  std::vector<ExecutionResult> results;
  for (const auto& statement : parsed.statements) {
    // The next bind occurs only after execute has successfully updated the real Catalog.
    results.push_back(executor.execute(compile_statement(statement, catalog, optimize).plan_after));
  }
  return results;
}
}  // namespace minisql
